### ISSUE TITLE
Generic return variable incorrectly resolves to `...[Unknown]`, only when bind

### ISSUE THREAD

**Issue by neutrinoceros** [astral-sh/ty#2796]
### Summary

Here's  somewhat reduced version of the actual code I'm type checking (it could probably be reduced further, though I'm hoping the problem is already evident):
```py
# t.py
from dataclasses import dataclass
from typing import Generic, TypeAlias, TypeVar, assert_type

import numpy as np

f32 = np.float32
f64 = np.float64
F = TypeVar("F", f32, f64)
FArray1D: TypeAlias = np.ndarray[tuple[int], np.dtype[F]]

@dataclass(kw_only=True, slots=True, frozen=True)
class Interval:
    min: float
    max: float

    def as_evenly_spaced_array(self, size: int, *, dtype: np.dtype[F]) -> FArray1D[F]:
        return np.linspace(self.min, self.max, size, dtype=dtype)

@dataclass(kw_only=True, slots=True, frozen=True)
class Grid(Generic[F]):
    x: FArray1D[F]
    y: FArray1D[F]

    @classmethod
    def from_intervals(
        cls,
        *,
        x: Interval,
        y: Interval,
        size: int,
        dtype: np.dtype[F],
    ) -> "Grid[F]":
        return Grid(
            x=x.as_evenly_spaced_array(size, dtype=dtype),
            y=y.as_evenly_spaced_array(size, dtype=dtype),
        )

def foo(x: FArray1D[F]) -> Grid[F]:
    return Grid.from_intervals(
        x=Interval(min=0.0, max=1.0),
        y=Interval(min=0.0, max=1.0),
        size=12,
        dtype=np.dtype("float64"),
    )

def bar(x: FArray1D[F]) -> Grid[F]:
    grid = Grid.from_intervals(
        x=Interval(min=0.0, max=1.0),
        y=Interval(min=0.0, max=1.0),
        size=12,
        dtype=np.dtype("float64"),
    )
    assert_type(grid, Grid[F])  # fails
    return grid  # pass
```

The *only* line for which a diagnostic is emitted is the second-to-last:
```
❯ ty check --python-version 3.11 src/lick/t.py
error[type-assertion-failure]: Argument does not have asserted type `Grid[F@bar]`
  --> src/lick/t.py:54:5
   |
52 |         dtype=np.dtype("float64"),
53 |     )
54 |     assert_type(grid, Grid[F])  # fails
   |     ^^^^^^^^^^^^----^^^^^^^^^^
   |                 |
   |                 Inferred type is `Grid[Unknown]`
55 |     return grid  # pass
   |
info: `Grid[F@bar]` and `Grid[Unknown]` are not equivalent types
info: rule `type-assertion-failure` is enabled by default

Found 1 diagnostic
```
This is highly surprising to me: I would expect the infered type of this variable to be identical to the function's return type *or* a diagnostic be raised also for the `return` statement.
In the actual code this is adapted from, there's buisiness logic instead of an `assert_type`, and I'm seeing what I interpret as false positives for `invalid-argument-type` and `invalid-return-type`. My understanding is that the type of `grid` being (incrorrectly) infered as `Grid[Unknown]` is the root cause.

### Version

ty 0.0.16 (e96ea4cd9 2026-02-10)

**Comment by MeGaGiGaGon** [comment]
This is happening because of the `np.dtype` calls, because the return type of `__new__` (#281) is still not respected, so all of [numpy's overloads](https://github.com/numpy/numpy/blob/0abf67d1785abfb4a2ad72fcbe7e00ab09f469b8/numpy/__init__.pyi#L1091) on the `__new__` of `dtype` refining the generic don't work, which means currently all calls to `np.dtype`'s `__new__` will give  `dtype[Unknown]/dtype[Any]`

**Comment by neutrinoceros** [comment]
It makes sense, but then why is there no diagnostics for return values ?

**Comment by MeGaGiGaGon** [comment]
Because `assert_type` is more strict that the return checks. `assert_type` means "exactly equal to", while returns just check for "assignable to". Because of the `Unknown` in `Grid[Unknown]` (which is treated like `Any`), `Grid[Unknown]` is assignable to any generic materialization of `Grid`. But `assert_type` does not care about how special `Any` is, since it only checks for an exact match.

**Comment by MeGaGiGaGon** [comment]
As for what's going wrong in your actual code (https://github.com/la-niche/lick/issues/327) I think it has something to do with ty's currently poor handling of constrained typevars, though I didn't look into it too deeply.

**Comment by neutrinoceros** [comment]
thanks a lot for the explanation !

> I think it has something to do with ty's currently poor handling of constrained typevars

most likely ! there's been a significant improvement in 0.0.16 (https://github.com/astral-sh/ty/issues/2728), and I'm trying to make sure the rest of the errors I'm seeing are known. If we're confident that this is merely an indirect effect of #281 then I'm down with closing this one.

**Comment by MeGaGiGaGon** [comment]
I think there is something else going on here too, since when I swap out the `np.dtype` for a class that does have a working generic type, the unknown still happens, so something should be wrong with ty's current handling of `from_intervals`
https://play.ty.dev/2b4aa0b5-f4aa-4827-adad-91cbbbb8cc7a
It may have to do with constrained typevars + classmethods? More investigation required.

Edit: It looks like it's related to using `F` in `from_intervals`, since swapping it out for an equivalent new typevar makes the later call make the expected `Grid[f64`
https://play.ty.dev/f3d9ebb3-60b7-45d8-a83b-acf274aa1fda

**Comment by neutrinoceros** [comment]
Thanks for digging in. I anticipate this issue's title would need editing once this Alaskan wolf is in sight.

**Comment by carljm** [comment]
Here's a minimal repro for the second issue (not about `__new__`) here:

```py
from typing import Generic, TypeVar

T = TypeVar("T")
U = TypeVar("U")

class Box(Generic[T]):
    @classmethod
    def create_same(cls, val: T) -> "Box[T]": ...

    @classmethod
    def create_diff(cls, val: U) -> "Box[U]": ...

reveal_type(Box.create_same(val=1))  # Box[Unknown]
reveal_type(Box.create_diff(val=1))  # Box[int]
```

The issue here is reuse of the `T` typevar, which is bound to the class, in the `create_same` classmethod. There is some disagreement about how this case should be handled. Mypy supports inferring T from this call; pyright and pyrefly do not. This (and the question of how we should handle it) is already covered in https://github.com/astral-sh/ty/issues/541

So I think this example is fully covered by #281 and #541; will somewhat-arbitrarily close it as duplicate of #541. Thanks for the report, and thanks @MeGaGiGaGon for the debugging work!

**Comment by neutrinoceros** [comment]
thanks ! will follow along these two !
